[GH-476] Allow specifying install method for windows_feature resources

dave-q opened the pull request:
- Introduces new attribute ['iis']['windows_feature_install_method']
if the node is present its value is passed into any windows_feature
resources as the install_method property

Signed-off-by: dave-q <24652224+dave-q@users.noreply.github.com>

# Description

Exposes the ability to dictate what install method is used by any windows_feature resources.


## Issues Resolved

[476](https://github.com/sous-chefs/iis/issues/476)

## Check List

- [x] All tests pass. See TESTING.md for details.
- [x] New functionality includes testing.
- [x] New functionality has been documented in the README if applicable.

Xorima commented on kitchen.yml:
```diff
@@ -71,3 +71,9 @@ suites:
       - recipe[test::vdir]
+  - name: default-windowsfeatures-powershell
+    run_list:
+      -recipe[iis::default]
```
you need a space in your yaml 
```suggestion
      - recipe[iis::default]
```

Xorima commented on recipes/mod_application_initialization.rb:
```diff
@@ -20,4 +20,7 @@
-windows_feature 'IIS-ApplicationInit'
+install_method = node['iis']['windows_feature_install_method']
+windows_feature 'IIS-ApplicationInit' do
+  install_method install_method unless install_method.nil?
```
If we are doing this everywhere I think we should just set the attribute to the current default in the attributes file?

dave-q replied:
my thinking here was to essentially always fallback to whatever windows_feature used as its default. That why if that resource implementation ever change, this cookbook would require no updates to match. I'm happy to not take that route though, as it would certainly make this change a lot less verbose.

Xorima commented on resources/install.rb:
```diff
@@ -21,6 +21,7 @@
 
 property :source, String
 property :additional_components, Array, default: []
+property :install_method, [:windows_feature_dism, :windows_feature_powershell, :windows_feature_servermanagercmd], required: false
```
You need to set a property type and then allowed symbols here, I think this is wrong. 

Best to look up Property types, https://docs.chef.io/custom_resources/#property (Which should be a symbol in this case) and validators to ensure it is equal to one of the allowed symbols you have set above: https://docs.chef.io/custom_resources/#validators

I would also set a default here so that the code does not need to change on line 34

Xorima commented on resources/manager.rb:
```diff
@@ -22,10 +22,12 @@
 property :enable_remote_management, [true, false], default: true
 property :log_directory, String
 property :port, Integer, default: 8172
+property :install_method, [:windows_feature_dism, :windows_feature_powershell, :windows_feature_servermanagercmd], required: false
```
See feedback for install resource.

Xorima commented on recipes/mod_aspnet.rb:
```diff
@@ -21,8 +21,11 @@
 include_recipe 'iis'
 include_recipe 'iis::mod_isapi'
 
+install_method = node['iis']['windows_feature_install_method']
```
Same feedback as for: mod_application_initialization.rb

Xorima commented on recipes/mod_aspnet45.rb:
```diff
@@ -21,4 +21,8 @@
 include_recipe 'iis::mod_isapi'
 
-windows_feature %w(NetFx4Extended-ASPNET45 IIS-NetFxExtensibility45 IIS-ASPNET45)
+install_method = node['iis']['windows_feature_install_method']
```
Same feedback as for: mod_application_initialization.rb

Xorima commented on recipes/mod_auth_basic.rb:
```diff
@@ -20,7 +20,11 @@
 include_recipe 'iis'
 
-windows_feature 'IIS-BasicAuthentication'
+install_method = node['iis']['windows_feature_install_method']
```
Same feedback as for: mod_application_initialization.rb

Xorima commented on recipes/mod_auth_digest.rb:
```diff
@@ -20,7 +20,11 @@
 include_recipe 'iis'
 
-windows_feature 'IIS-DigestAuthentication'
+install_method = node['iis']['windows_feature_install_method']
```
Same feedback as for: mod_application_initialization.rb

Xorima commented on recipes/mod_auth_windows.rb:
```diff
@@ -20,7 +20,11 @@
 include_recipe 'iis'
 
-windows_feature 'IIS-WindowsAuthentication'
+install_method = node['iis']['windows_feature_install_method']
```
Same feedback as for: mod_application_initialization.rb

Xorima commented on recipes/mod_cgi.rb:
```diff
@@ -20,4 +20,8 @@
 include_recipe 'iis'
 
-windows_feature 'IIS-CGI'
+install_method = node['iis']['windows_feature_install_method']
```
Same feedback as for: mod_application_initialization.rb

Xorima commented on recipes/mod_compress_dynamic.rb:
```diff
@@ -20,4 +20,8 @@
 include_recipe 'iis'
 
-windows_feature 'IIS-HttpCompressionDynamic'
+install_method = node['iis']['windows_feature_install_method']
```
Same feedback as for: mod_application_initialization.rb

Xorima commented on recipes/mod_compress_static.rb:
```diff
@@ -20,4 +20,8 @@
 include_recipe 'iis'
 
-windows_feature 'IIS-HttpCompressionStatic'
+install_method = node['iis']['windows_feature_install_method']
```
Same feedback as for: mod_application_initialization.rb

Xorima commented on recipes/mod_ftp.rb:
```diff
@@ -20,4 +20,8 @@
 include_recipe 'iis'
 
-windows_feature %w(IIS-FTPServer IIS-FTPSvc IIS-FTPExtensibility)
+install_method = node['iis']['windows_feature_install_method']
```
Same feedback as for: mod_application_initialization.rb

Xorima commented on recipes/mod_iis6_metabase_compat.rb:
```diff
@@ -20,4 +20,8 @@
 include_recipe 'iis'
 
-windows_feature %w(IIS-IIS6ManagementCompatibility IIS-Metabase)
+install_method = node['iis']['windows_feature_install_method']
```
Same feedback as for: mod_application_initialization.rb

Xorima commented on recipes/mod_isapi.rb:
```diff
@@ -20,4 +20,8 @@
 include_recipe 'iis'
 
-windows_feature %w(IIS-ISAPIFilter IIS-ISAPIExtensions)
+install_method = node['iis']['windows_feature_install_method']
```
Same feedback as for: mod_application_initialization.rb

Xorima commented on recipes/mod_logging.rb:
```diff
@@ -20,4 +20,8 @@
 include_recipe 'iis'
 
-windows_feature 'IIS-CustomLogging'
+install_method = node['iis']['windows_feature_install_method']
```
Same feedback as for: mod_application_initialization.rb

Xorima commented on recipes/mod_management.rb:
```diff
@@ -20,7 +20,10 @@
 
 include_recipe 'iis'
 
+install_method = node['iis']['windows_feature_install_method']
```
Same feedback as for: mod_application_initialization.rb

Xorima commented on recipes/mod_security.rb:
```diff
@@ -20,4 +20,8 @@
 include_recipe 'iis'
 
-windows_feature %w(IIS-URLAuthorization IIS-RequestFiltering IIS-IPSecurity)
+install_method = node['iis']['windows_feature_install_method']
```
Same feedback as for: mod_application_initialization.rb

Xorima commented on recipes/mod_tracing.rb:
```diff
@@ -20,4 +20,8 @@
 include_recipe 'iis'
 
-windows_feature 'IIS-HttpTracing'
+install_method = node['iis']['windows_feature_install_method']
```
Same feedback as for: mod_application_initialization.rb

Xorima commented:
@dave-q please can you also add an entry in the changelog file under the `## Unreleased` header

dave-q commented on kitchen.yml:
```diff
@@ -71,3 +71,9 @@ suites:
+      - recipe[iis::default]
+    attributes:
+      iis:
+        windows_feature_install_method: :windows_feature_powershell
```
it appears that this doesn't translate to a ruby symbol :(

dave-q commented:
@Xorima So after a bit of fiddling I've got the test's passing. The code in the PR is all working but in my opinion I need to give it one more pass over and tidy it up. But I was hoping to get your opinion on a couple of things.

1. I've added this helper function to translate 'dism-format' feature names into 'powershell-format', if the install method is powershell - While this works pretty well and saves having to have a conditional in each place that wants a feature managed (to decide which name to use). It also makes the api more flexible, it can accept the additional features node containing either format. However one could argue that its extra complexity and in a way the fact that the 'api' is so 'forgiving' could infact be a little confusing.. I'd be interested to hear your's (or anyone elses) opinion on this. The other option is that all the recipes that install windows features decide which format to use in the recipe and then use that with windows_feature
2. If we think the that 'translating' the feature name depending on the install method is good then that leaves me with one more questions. Currently I've introduce a helper method in a library AND a wrapper resource. I think it should be one or the other. Either a library that is called in every recipe, that then goes on to called windows_feature. Or we continue to have the wrapper resource and the resource can just define the functionality for 'feature name translation'. Again I'm not sure whats better. I'd lean towards the wrapper resource.. but that maybe due to my background of OOP...

Let me know your thoughts and I can update the PR

Xorima commented:
@rlaveycal  would you be able to take a look over these changes as a heavy user of the iis cookbook?

rlaveycal commented:
I'd lean more to the library-in-resource and use `window_feature` approach because it avoids having to keep the wrapper resource in API sync. It also avoids the temptation for people to use this resource themselves.

bmhughes commented:
Definetly go with the library method route imo. 

The conversion helper method can be refactored to make it more idiomatic ruby wise:
1. Use `map` instead of each which will shorten it and remove several intermediate variables
2. Instead of using splat to pass the feature names in it would be better served by passing an Array to a single parameter. Splat is idomatically used to make a variable number of different parameters into a method, as this takes a single parameter (that may be of variable length) it doesn't really make sense to use splat. Wrap it in `Array()` to cover the cases where just a string is passed in.

dave-q commented:
@bmhughes @rlaveycal 

Thanks for taking the time to review my PR. I've updated it with your suggestions and all the tests are passing.

I have a couple of questions.

1. I wasn't sure about Monkey patching the ::Chef::DSL::Recipe class with my helper, but I was told and from doing some reading too, this is the correct way to do this, anyway.. I'm happy if you guys are but if you have a suggestion for a different way, I'm all ears
2. My initial attempt was incorrect (i just called include IISCookbook::WindowsFeatureHelper directly in the recipe) however I didn't notice that until the site inspec test ran... I would have caught it much sooner, if there was just a basic 'it converges' spec test for all recipes. I'm happy to add them if you think its worth it.

Thanks

bmhughes commented on resources/install.rb:
```diff
@@ -18,17 +18,20 @@
 
 property :source, String
 property :additional_components, Array, default: []
+property :install_method, Symbol, required: false, equal_to: [:windows_feature_dism, :windows_feature_powershell, :windows_feature_servermanagercmd], default: :windows_feature_dism
```
Rather than having the multiple `to_sym` calls in various places, change this property type to `[String, Symbol]` and add a coerce of `proc { |p| p.to_sym}` to the property definition so you'll always have a symbol further on down the call stack. It also opens this up to greater flexibility where passing a symbol isn't as easy as it should be (instantiating from roles where symbol doesn't exist in json for one).

lamont-granquist replied:
yep

dave-q replied:
I can certainly do this for the install resource. However 99% of the `tosym` calls are for calling into the windows_feature resource directly from a recipe, so I don't have to call it there, since we are taking the value directly from the attribute.
if we expect the attribute to be a symbol, it won't work in test kitchen and we can't test it properly

bmhughes commented:
> 
> 
> @bmhughes @rlaveycal
> 
> Thanks for taking the time to review my PR. I've updated it with your suggestions and all the tests are passing.
> 
> I have a couple of questions.
> 
>     1. I wasn't sure about Monkey patching the ::Chef::DSL::Recipe class with my helper, but I was told and from doing some reading too, this is the correct way to do this, anyway.. I'm happy if you guys are but if you have a suggestion for a different way, I'm all ears
> 
>     2. My initial attempt was incorrect (i just called include IISCookbook::WindowsFeatureHelper directly in the recipe) however I didn't notice that until the site inspec test ran... I would have caught it much sooner, if there was just a basic 'it converges' spec test for all recipes. I'm happy to add them if you think its worth it.
> 
> 
> Thanks

You've got a couple of options here depending on **where** you need the helper method to be available. From a quick scan through I can't see you needing it outside of the custom resource so I wouldn't mix it into the recipe DSL class for this, just include it in the outer resource definition (which I think you already have done) and drop the include to the recipe class.

Something to note is that with how it is now the method will only be mixed into the outer resource definition class (properties etc) and **won't** be in the action_class(es), if you needed the module methods within the action class(es) as well then that's where the `action_class` block comes in handy as each action class will `yield` that on compiliation and include the module within all the actions of the resource.

lamont-granquist commented on libraries/windowsfeature_helper.rb:
```diff
@@ -0,0 +1,42 @@
+
+module IISCookbook
+  module WindowsFeatureHelper
+    def transform_feature_name(install_method, names)
```
the `install_method` is tightly coupled to the implementation of the rest of your code so this shouldn't be exposed as a public API in the recipe class.

in fact since its only used in one place it isn't really appropriate to extract it as a helper at all, so it should live with the code that calls it in the resource in a action_class declaration.

lamont-granquist replied:
it is also pretty weird to have the helper take the install_method and then return the names arg unmatched.

i'd say this should probably be renamed `powershell_feature_name(names)` and the check moved to the caller.

then at some later date if this class actually grows the need for multiple differen kinds of feature_name filters then you can write a method with this signature and switch/case on the install_method to transform them.

dave-q commented:
@lamont-granquist @bmhughes 
I appreciate the feedback. You've both made a couple of comments that seemed to be based around the thought that the only place we are using the new attribute (windows_feature_install_method) is in the install.rb resource. It is used in almost every recipe (any starting with mod_) since all those recipes use the windows_feature resource, so we need to pass that into those resources.

We also kind of need to use that Module helper method I added, in all those mod_* recipes since we need to translate the feature name. We could do a switch on the install mode and 'hardcode' the powershell version of feature name in there too

something like 

```
feature_name = ''
if install_method == :windows_powershell do
  feature_name = <powershell format>
else
  feature_name = <dism format>
end
```

personally I think thats a little messy, but I can see an argument for doing it

anyway, I'm going to hold off making any changes right now just until this is a bit clearer

dave-q commented on recipes/mod_application_initialization.rb:
```diff
@@ -19,5 +19,8 @@
 #
 
 include_recipe 'iis'
+install_method = node['iis']['windows_feature_install_method']&.to_sym
```
heres an example of where the node attribute is used and on line 24 where the tranform_feature_name method is used.

All the recipes starting with mod_ are the same

bmhughes commented:
I know the sticks in the mud hate it but is there any reason those recipes can't go away and be replaced by a resource? A lot of them just seem to be doing the same thing from a quick skim through.

dave-q commented:
> I know the sticks in the mud hate it but is there any reason those recipes can't go away and be replaced by a resource? A lot of them just seem to be doing the same thing from a quick skim through.

I've got no strong opinion one way or the other really, but just so I'm clear what you are suggesting.

The recipes wouldn't exist at all? And people should call them as resources?? That would definitely be a big breaking change wouldn't it? Since anyway who utilised those recipes in a downstream cookbook would get their cookbook broken right?

If the recipes still existed..they would just be calling in to a resource.. much like 90% of them do now anyway. It would just be a resource from this cookbook rather than a core resource. I did take a stab at that in one of my earlier iterations, essentially a windows_feature wrapper that did all of the stuff with install modes and symbols and feature names. But we opted to move away from that as it would be exposed as a resource for others to use and also we'd have to keep the API in sync. I'm happy to put that back on the table if you think that's better.

bmhughes commented:
Yes that's what I meant (I'd wait for someone else to chime in incase I'm missing something obvious though), apologies for the garden path route to it.

The massive breaking change is why I made the sticks in the mud comment, in general we're (sous-chefs) trying to move to resource-only cookbooks where the final implementation is created by the user in a wrapper cookbook that wraps the resources to what _they_ need rather than trying to cover 90% of use cases with a generic recipe. Some people are aginst this because they have to create org specific wrapper cookbooks, but with the right style and some forethough they're very quick to create and have very little substance (the guts of the logic are implemented by the resources) so aren't the end of the world. It's the same paradigm as terraform/ansible when you add a provider or module.

I wouldn't create (or think of it as) a wrapper for the windows_feature though, I (personally) would call it module_install or something along those lines. The fact that is will mostly be a wrapper around windows_feature is irrelevant as the desired outcome of the resource is to install an IIS module, it just happens to use windows_feature to achieve that in the same way as a lot of the other resources in sous cookbooks use package, template etc. The core resources don't undergo massive API changes in general without giving you plenty of notice and warning, so I don't see that as an issue.

That's my 2p anyway.  I reserve the right to be completely wrong 😂.

dave-q commented:
I can definitely see the benefits of that model for sure.
I'm to familiar with the use cases of chef. I know at my work we have our own internal supermarket so creating a wrapper would be simple enough. But what about companies that don't have that infrastructure and just want to use public recipes directly with some other orchestrating framework? I suppose I'm just speculating now really.

Anyway... For this PR, would a good compromise be to introduce a new resource 'module_install' (much better name) that wraps all that logic nicely and then for now update the recipes to just call that resource so they essentially stay the same but just call the new resource.

Then in the future it would be simple to remove the recipes in a larger breaking change release?

bmhughes commented:
I'm open to the general concensus on that, personally I'd just rip it all out including the attributes and move everything to library/resources. But it certainly is an alternative option and does deal with removing the need to mix that library (or for it to exist at all as @lamont-granquist points out) into the recipe DSL unnecessarily.

dave-q commented:
So I've taken a lot of what you both suggested on board @bmhughes @lamont-granquist 

I noticed that with a little refactoring, the install resource could do all the new work translating feature names and optionally do the starting of the W3SVC service if desired.

This makes the recipes really really basic now, they all just call into the same resource (some call other internal resources too)

If we want to take the route @bmhughes  said of removing the recipes, this would be very easy in the future.

I added some new spec tests for the install resource since it had a little more 'behavior' in it.

I also added a spec test for all the mod_* recipes. I think its a good habit and gives some quick confidence that things aren't broken.

Look forward to hearing your thoughts

Thanks

Xorima commented on README.md:
```diff
@@ -36,6 +36,7 @@ This cookbook is maintained by the Sous Chefs. The Sous Chefs are a community of
 - `node['iis']['pubroot']` - . default is `%SYSTEMDRIVE%\inetpub`
 - `node['iis']['docroot']` - IIS web site home directory. default is `%SYSTEMDRIVE%\inetpub\wwwroot`
 - `node['iis']['cache_dir']` - location of cached data. default is `%SYSTEMDRIVE%\inetpub\temp`
+- `node['iis']['windows_feature_install_method']` - specify the install method that will be used by any windows_feature resources. If ommitted it will not be specified and will the windows_feature supply a default. Valid options are ':windows_feature_dism, :windows_feature_powershell, :windows_feature_servermanagercmd'. Default is :windows_feature_dism
```
```suggestion
- `node['iis']['windows_feature_install_method']` - specify the install method that will be used by any windows_feature resources. If ommitted it will not be specified and will use `windows_feature_dism` by default. Valid options are `:windows_feature_dism`, `:windows_feature_powershell`, `:windows_feature_servermanagercmd`. Default is `:windows_feature_dism`
```

Xorima commented on resources/install.rb:
```diff
@@ -20,15 +20,54 @@
 
 property :source, String
-property :additional_components, Array, default: []
+property :additional_components, [Array, String],
```
why should this now accept a string?

dave-q replied:
If you only want to specify one additional feature, as many recipes do, it saved cluttering every recipe with code to make a single item array

Xorima commented on resources/install.rb:
```diff
@@ -20,15 +20,54 @@
+  coerce: proc { |i| i.to_sym },
+  equal_to: [:windows_feature_dism, :windows_feature_powershell, :windows_feature_servermanagercmd],
+  default: :windows_feature_dism
+property :start_iis, [true, false], default: false
```
Why are we controlling the service in the install resource now?

dave-q replied:
Every recipe in the cookbook was doing the same things really.

Include the default recipe, add another windows feature and then sometimes one other resource.

The default recipe would install iis and then start it.

Moving the starting into the resource (bare in mind it's optional and doesn't by default to keep past behaviour the same for users of the resource) meant that 

1. Every recipe could now call the install resource and pass its own additional features along with, removing the need to use windows_feature in almost every recipe.
2. Now there was only one place windows_feature is being used. The install resource, this allows us to keep the 'translation of feature names' just contained to that resource.
3. All the recipes are dead simple now and if at some point in the future sous-chefs want to remove them it will be much easier for people to transition and keep the same behaviour

bmhughes replied:
It's been a while since i've had to use IIS but i'm pretty sure the w3svc service is enabled by default when IIS is installed? So I don't think this is necessary. Also, it's adding unexpected functionality to the resource, I don't expect an install resource to do anything other than install.

If people need to manage the IIS service then it's not a big deal to do that in the implementing wrapper as there's no abstraction required outside of the core Chef resources.

dave-q replied:
It doesn't start it by default you have to set the property.
The reason I added it was it allows the rest of the recipes to be much simpler, since they all called the default recipe anyway and the default recipe started the service and I wanted to ensure I wasn't changed anything for existing people who don't want to have to update how they use this cookbook. It means that all recipes rely on the install resource to install their features and we can keep that feature name translation in one place rather than having to spread it around the cookbook.

For outside users the install resource really hasn't changed if they don't want. By default it won't try and start the service (which according yourself is probably done anyway), personally I think this gives us the benefit of new functionality with out breaking any existing features/APIs, whilst also laying the ground work for the future desire of just exposing resources rather than these recipes. The recipes are now very basic so it's easy for people to 're-implement' them in any future breaking change

Xorima commented on resources/install.rb:
```diff
@@ -20,15 +20,54 @@
+      if result.stderr.to_s.empty?
+        next result.stdout.strip
+      else
+        Chef::Log.warn("Unable to translate feature #{name}")
```
Should this be a silent failure? I think if given an incorrect name it should throw an error else people who do not read the logs may have a different expectation compared to reality

damacus commented:
@Xorima and @bmhughes are you both happy with the way this current PR is?

kitchen-porter commented:
Released as: [8.1.0](https://github.com/sous-chefs/iis/releases/tag/8.1.0)
